### ISSUE TITLE
MariaDB: Database backup/restore after upgrade?

### ISSUE THREAD

**Issue by klonos** [WPN-XM/WPN-XM#361]
Trying to restore my old databases by copying the ```\bin\mariadb\data``` directory over to the new installation. This might not be the proper way, but I was going for something "quick 'n' dirty". Here's the error log:

```
...
2015-03-16 01:28:50 8524 InnoDB: Error: Fetch of persistent statistics requested for table "labquipsales"."search_index" but the required system tables mysql.innodb_table_stats and mysql.innodb_index_stats are not present or have unexpected structure. Using transient stats instead.
2015-03-16 01:28:53 4254 InnoDB: Error: Column last_update in table "mysql"."innodb_table_stats" is INT UNSIGNED NOT NULL but should be BINARY(4) NOT NULL (type mismatch).
...
150316  1:30:23 [Note] feedback plugin: report to 'https://mariadb.org/feedback_plugin/post' was sent
150316  1:30:23 [Note] feedback plugin: server replied 'ok'
```

**Comment by klonos** [comment]
The two ```Using transient stats instead.``` and ```type mismatch``` error were repeated dozens of times. The last two line seem to indicate that a report was sent back to the "mothership". So I guess not really a WPN-XM issue, but wanted to document this for others that might be having the same issue.

**Comment by klonos** [comment]
...well if this is a MariaDB issue and it ends up being fixed in a newer version, perhaps it will as well be time for a new minor version of WPN-XM that includes that version.

**Comment by jakoch** [comment]
With WPN-XM v0.8.2 you should have v10.1.3 (x86/x64) of MariaDb.
Which issue exactly? Is it in their bugtracker?

Yes, feeback mode is on - i discussed this with the Maria guys. They need usage stats and want to know the usage and distributions of features and versions on Windows.

- [ ] we need proper instructions how to save and restore a database (wiki or docs)

**Comment by klonos** [comment]
Documentation is gold, but it would be even better if we could automate this and offer in-place upgrades.

**Comment by jakoch** [comment]
Physical backup of the files will not work for different versions.
An automation could use a logical database dump or export, maybe via mysqldump and then import that, after a db version upgrade. 
Dumping could also be part of the backup system.

**Comment by klonos** [comment]
I was wondering... when I upgrade the MariaDB package on my actual production Ubuntu servers, there's nothing that I need to do - the new version is simply installed over the old one and db files are left there intact. Simply works. Are you sure that the whole dump import/export thing is necessary? Would perhaps simply not deleting the ```\bin\mariadb\data``` folder suffice?

**Comment by jakoch** [comment]
> Are you sure that the whole dump import/export thing is necessary?

No, i'm not sure. It's just an suggestion on how an upgrade mechanism could work.
But i know from the MariaDb manual, that only an service upgrade is easy and automated and the  non-service upgrade is a complicated multiple-step manual process.

How do other stacks, like EasyPHP/WampServer or XAMPP handle this?

>Would perhaps simply not deleting the \bin\mariadb\data folder suffice?

Deleting the `\data` folder first, means that it's recreated because of the `mysql_install_db` execution - and that's actually like a fresh install.

--

We need some checks here:
- which important files are overwritten during the extraction on top of an existing install
- what happens, if its an upgrade and i leave the execution of `mysql_install_db` out, see https://github.com/WPN-XM/WPN-XM/blob/master/installers/full-php55-w32.iss#L1285-L1289
    
It could be that the call to `mysql_install_db` is the culprit, because an already initialized database is initialized again and thereby possibly resetted. Needs testing...

**Comment by jakoch** [comment]
> The purpose of the mysql_install_db program is to initialize the data directory, including the tables in the mysql system database. It does not overwrite existing MySQL privilege tables, and it does not affect any other data.

This is not explicitly stated in the Maria docs https://mariadb.com/kb/en/mariadb/mysql_install_db/
but in http://dev.mysql.com/doc/refman/5.7/en/mysql-install-db-problems.html
`mysql_install_db` is not the culprit - according to the docs.

**Comment by jakoch** [comment]
Bounty: 600 €.
